inference-engine/src/inference_engine/src/cpp/async_infer_queue.cpp

```diff
@@ -0,0 +1,135 @@
+// Copyright (C) 2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "openvino/runtime/async_infer_queue.hpp"
+
+#include <condition_variable>
+#include <map>
+#include <mutex>
+#include <queue>
+#include <string>
+#include <vector>
+
+#include "ie/ie_plugin_config.hpp"
+
+namespace ov {
+namespace runtime {
+
+AsyncInferQueue::AsyncInferQueue(ov::runtime::ExecutableNetwork net, size_t jobs) {
+    // Automatically set number of jobs
+    if (jobs <= 0) {
+        try {
+            auto parameter_value = net.get_metric(METRIC_KEY(SUPPORTED_METRICS));
+            auto supported_metrics = parameter_value.as<std::vector<std::string>>();
+            const std::string key = METRIC_KEY(OPTIMAL_NUMBER_OF_INFER_REQUESTS);
+            if (std::find(supported_metrics.begin(), supported_metrics.end(), key) != supported_metrics.end()) {
+                parameter_value = net.get_metric(key);
+                if (parameter_value.is<unsigned int>())
+                    jobs = parameter_value.as<unsigned int>();
+                else
+                    IE_THROW() << "Unsupported format for " << key << "!"
+                               << " Please specify number of infer requests directly!";
```

Comment:
These error messages are a little misleading. The network is not loaded in this constructor, the queue object's "pool" is configured here. Can you adjust them? Also instead of `directly` I suggest using `explicitly`.